### ISSUE TITLE
How to deal with filegroups?

### ISSUE THREAD

**Issue by lipe123** [olahallengren/sql-server-maintenance-solution#260]
I can't find any documentation on how to backup filegroups individually as-is instead of combining them back into one large database backup.

I have two 300+gb databases (one live one test) and the regular backup script keeps failing. My only guess is some kind of timeout or similar issue with the large transfer taking very long.

I'd love to be able to backup the individual 10 filegroup files somehow to cut down on the single large transfer time.

Any suggestions/help please?

**Comment by JonasNjopOlsson** [comment]
You could check my pull request which added filegroup backup support to the solution: #217 

I will later this month update the pull request so that it doesn't conflict with the current main branch. We are using my code in production and it is working well for the very large databases that we have with read-only filegroups.

**Comment by lipe123** [comment]
Hi, thanks for the heads up. I'm pretty super clueless still with all this. 
Looking at the issue 38 and pull info it looks like it would still backup the filegroup as a single backup file or does it backup each file in the filegroup separately?

**Comment by JonasNjopOlsson** [comment]
Yes, it will back up one filegroup at a time, not the individual files in the database. I thought that was what you were asking for? It is certainly possible to back up invidiual logical files instead, using the FILE option to the BACKUP command but that is a bit different from backing up using the FILEGROUP option.

Are you saying that you essentially want to create a snapshot copy of the MDF, NDF and LDF files in a SQL Server database instead of creating backup files from them? I don't think that Hallengren's solution can help with that though as that would require some support on the filesystem level and below. That will require some third party vendor support, I believe.

Also, think of how to perform restores when doing backups of separate parts of a database. You can find more information here:

- https://docs.microsoft.com/en-us/sql/relational-databases/backup-restore/piecemeal-restores-sql-server
- https://www.sqlshack.com/database-filegroups-and-piecemeal-restores-in-sql-server/

If I were you, I would look into the real issue which is the failing network data transfers that you are seeing. Don't try to get around that by complicating your backup and restore routine.

Hallengren's script has support for backing up a database to multiple backup files, splitting the contents between them. This usually speeds up backups quite a bit in the environments I have seen:

- https://jackworthen.com/2017/12/11/backing-up-a-sql-server-database-into-multiple-backup-files/
- https://www.brentozar.com/archive/2014/01/improving-the-performance-of-backups/

**Comment by olahallengren** [comment]
Here’s a test version of DatabaseBackup with support for filegroup-level backup:
https://github.com/olahallengren/sql-server-maintenance-solution/tree/file_group_backup

I’m looking for early testers to help validate the functionality.
If you’re interested, feel free to reach out to me via [email](https://ola.hallengren.com/contact.html)

**Comment by olahallengren** [comment]
I've dropped the `file_group_backup` branch for now. I've been working in other areas of `DatabaseBackup` and the branch fell behind, but I plan to get back to it.
